### PR TITLE
Small fixes for convinient work with 2d data; add paired image datasets

### PR DESCRIPTION
**Pull request by ocenandor**
1. Change `sample2d.py` to have an oportunity to save raw generated points with `.th` extention 
2. `train2d.py`
    * remove integrator arguments 
    * set default name of experiment - `datetime.now().strftime("%y-%m-%d %H:%M:%S")`
    * checkpoint dir changed to `log_dir / name`
    * add `--ckpt_every` arg to save multiple checkpoints
3. Add multiple checkpointing to `utils.py`
4. Apply isort
5. `train.py`
    * set default name of experiment - `datetime.now().strftime("%y-%m-%d %H:%M:%S")`
    * checkpoint dir changed to `log_dir / name`
    * add `--ckpt_every` arg to save multiple checkpoints
    * move corruption funciton from main
    * change args to dataset1 (target), and dataset2 (initial, optional)
6. Add paired dataset in `dataset.py` (update `get_pair_dataset`, a bit `get_dataset`)
7. Add quick start to `README.md`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,16 @@
 # toySB
 Schrodinger Bridge realization for toy datasets
 
+# Quick start 🚀
+
+```bash
+git clone https://github.com/David-cripto/toySB.git
+cd toySB
+git submodule update --init --recursive
+pip install -e .
```
Comment (David-cripto) [inline README.md:10]: -e - editable mode, need only for developers

File: scripts/sampling/sample2d.py [hunk 2]
```diff
@@ -22,6 +22,7 @@ def create_arguments():
     parser.add_argument("--ema", type=float, default=0.99)
     parser.add_argument("--ab_order", type=int, default=0, help="order of polynom in Exponential Integrator")
     parser.add_argument("--nfe", type=int, default=1000, help="number of function evaluations")
+    parser.add_argument("--save_raw_data", action="store_true", help="save raw data (not only plots)")
```
Comment (David-cripto) [inline scripts/sampling/sample2d.py:25]: What is it?

File: scripts/sampling/sample2d.py [hunk 3]
```diff
@@ -41,7 +42,10 @@ def main(opt):
 
     val_dataloader = DataLoader(val_dataset, len(val_dataset))
     
-    sampling(opt, val_dataloader, net, ema, scheduler, opt.path_to_save)
+    experiment_name = Path(opt.ckpt_path).parts[-2]
```
Comment (David-cripto) [inline scripts/sampling/sample2d.py:45]: why -2?

File: scripts/train/train2d.py [hunk 4]
```diff
@@ -31,20 +36,17 @@ def create_arguments():
...
     parser.add_argument("--ot_ode", action="store_true", help="use OT-ODE")
-    parser.add_argument("--exp_int", action="store_true", help="use Exponential Integrator")
     parser.add_argument("--nfe", type=int, default=1000, help="number of function evaluations")
-    parser.add_argument("--ab_order", type=int, default=0, help="order of polynom in Exponential Integrator")
 
     opt = parser.parse_args()
 
     opt.device='cuda' if opt.gpu is None else f'cuda:{opt.gpu}'
     os.makedirs(opt.log_dir, exist_ok=True)
-    os.makedirs(opt.ckpt_path, exist_ok=True)
```
Comment (David-cripto) [inline scripts/train/train2d.py:42]: Why did you delete it?

File: requirements.txt [hunk 5]
```diff
@@ -6,4 +6,5 @@ diffusers
 tqdm
 jax
 torch-ema
-taming-transformers
\ No newline at end of file
+taming-transformers
+ipdb
```
Comment (David-cripto) [inline requirements.txt:10]: We should add version of the libs. I was too lazy to do it, but we need to do it

File: scripts/train/train.py [hunk 6]
```diff
@@ -5,20 +5,24 @@
...
     parser.add_argument("--root", type=Path, default="", help="path to save data")
     parser.add_argument("--log-dir", type=Path, default=".log", help="path to log std outputs and writer data")
     parser.add_argument("--ckpt_path", type=Path, default="", help="path to save checkpoints")
+    parser.add_argument("--ckpt_every", type=int, default=0, help="period of checkpointing; 0 - save only last")
     parser.add_argument("--gpu", type=int, default=None, help="choose a particular device")
     parser.add_argument("--num_steps", type=int, default=1000, help="number of steps")
     parser.add_argument("--num_epoch", type=int, default=100, help="number of epochs")
     parser.add_argument("--image_size", type=int, default=28, help="image size")
-    parser.add_argument("--c_in", type=int, default=3, help="in channels")
+    parser.add_argument("--c_in", type=int, default=3, help="in channels") # TODO better to set 3 and forget about this
```
Comment (David-cripto) [inline scripts/train/train.py:25]: Mb, you're right. I just was thinking about expanding this to other domains(not only for pictures)

File: scripts/train/train.py [hunk 7]
```diff
@@ -35,29 +39,29 @@ def create_arguments():
...
-    corruption_func = build_sr4x(opt, logger, "bicubic", opt.image_size)
-    train_pair_dataset = get_pair_dataset(opt, logger, corruption_func, train=True)
-    val_pair_dataset = get_pair_dataset(opt, logger, corruption_func, train=False)
+    train_pair_dataset = get_pair_dataset(opt, logger, train=True)
+    val_pair_dataset = get_pair_dataset(opt, logger, train=False)
     net = get_model(image_size = opt.image_size, in_channels = opt.c_in, num_channels = 64, num_res_blocks = 5)
     scheduler = Scheduler(create_symmetric_beta_schedule(n_timestep=opt.num_steps, linear_end=opt.beta_max / opt.num_steps), opt.device)
     train_dataloader = DataLoader(train_pair_dataset, opt.batch_size, shuffle = True)
-    val_dataloader = DataLoader(val_pair_dataset, opt.val_log)
+    val_dataloader = DataLoader(val_pair_dataset, opt.val_log, shuffle = True)
```
Comment (David-cripto) [inline scripts/train/train.py:64]: It's bad to shuffle validation dataloader, 'cause order here is important for logging images

File: toysb/utils.py [hunk 8]
```diff
@@ -250,17 +257,24 @@ def train(opt, net, scheduler, train_dataloader, val_dataloader, logger):
...
+        "ema": ema.state_dict(),
+        "optimizer": optimizer.state_dict(),
+        "sched": sched.state_dict() if sched is not None else sched,
+        }, opt.log_dir / opt.name / f"step=last_loss={loss.item():.2f}.pt")
+    logger.info(f"Saved latest({it=}) checkpoint to {(opt.log_dir / opt.name)=}!")
     writer.close()
 
 def load_from_ckpt(net, opt, logger):
-    checkpoint = th.load(opt.ckpt_path, map_location="cpu")
+    checkpoint = th.load(opt.ckpt_path, map_location="cpu") #FIXME possible bug with ckpt_path
```
Comment (David-cripto) [inline toysb/utils.py:277]: What's kind of the bug here?

File: toysb/utils.py [hunk 9]
```diff
@@ -202,12 +205,16 @@ def pred_x0_fn(xt, step):
         else:
             save_imgs(xs, log_steps, path_to_save)
     
+    try: #FIXME Do union argparser for sample and train, and put in in  separate file (i.e. utils)
+        if opt.save_raw_data:
+            th.save({"xs": xs, "x0": x0}, path_to_save / '../raw_data.pt')
+    except AttributeError:
+        pass
```
Comment (David-cripto) [inline toysb/utils.py:212]: Why do you need to save it?

**Review by David-cripto — CHANGES_REQUESTED** [review]
Some changes are not appropriate for merging. Overallб I like most of the changes

File: scripts/train/train2d.py (hunk 4 above)
Comment (ocenandor) [reply scripts/train/train2d.py:42]: I propose the following structure of loggi(ng files
```
.log # from --log-dir
  |  — experiment_name # from --name
     |  — tensorboard_logs.tfevents
     |  — ckpt_1.pt
     | ...
     |  — ckpt_n.pt
```

Also change code in `sampling.py` for convinient argumebts setting according ti this structure

So, there is no need to create a ckpt_path

File: toysb/utils.py (hunk 8 above)
Comment (ocenandor) [reply toysb/utils.py:277]: Mark this, because don't manually check this function after changing arguments in train and sample

File: toysb/utils.py (hunk 9 above)
Comment (ocenandor) [reply toysb/utils.py:212]: drawing trajectories

File: scripts/sampling/sample2d.py (hunk 2 above)
Comment (ocenandor) [reply scripts/sampling/sample2d.py:25]: For drawing trajectories

File: toysb/utils.py (hunk 8 above)
Comment (ocenandor) [reply toysb/utils.py:277]: But it works, everything's fine

File: scripts/sampling/sample2d.py (hunk 3 above)
Comment (ocenandor) [reply scripts/sampling/sample2d.py:45]: Log folder is formed according to the naming used in train procedure

File: toysb/datasets/dataset.py
[the remark this answers is not in the document]
Comment (ocenandor) [reply toysb/datasets/dataset.py:1]: `PAIRED_MODE = True` if we use setups like mnist <-> emnist

For super-resolution task `PAIRED_MODE = False` and the the old logic and code is used

```
self.indexes = [
    get_correspondance_index(name) for name in [opt.dataset1, opt.dataset2]
]
```
Here (lines 68-70) get bicection between labels of different datasets (0 = A, 1 = S, 2 = F, etc. fpr emnist <-> mnist)

Further (71-74) get bicection between samples of corresponding labels

`self.raw_idx` (76) - just usual index used for __getitem __

**Review by David-cripto — APPROVED** [review]
